File: desktop/src/support/webcore/icons.cpp
#include "icons.hpp"

#include "skinPrefs.hpp"

#include <QFile>
#include <QHash>
#include <QJsonArray>
#include <QJsonDocument>
#include <QJsonObject>
#include <QPainter>
#include <QStringList>

static void ensureAppResources() { Q_INIT_RESOURCE(app); }

namespace stencil::support {

  namespace {
    struct PixelSet {
      int size = 16;
      QChar logoFrame;                 // the mark's ring: the one cell the accent inks
      QHash<QChar, QString> palette;   // a char → its hex; the dot is absent
      QHash<QChar, QString> dark;      // the characters the dark face re-inks
      QHash<QString, QStringList> icons;
    };

    QHash<QChar, QString> readPalette(const QJsonObject& o) {
      QHash<QChar, QString> out;
      for (auto it = o.begin(); it != o.end(); ++it)
        if (!it.value().isNull() && !it.key().isEmpty()) out.insert(it.key().at(0), it.value().toString());
      return out;
    }

    const PixelSet& pixelSet() {
      static const PixelSet set = [] {
        ensureAppResources();
        PixelSet s;
        QFile f(":/config/iconsWebcore.json");
        if (!f.open(QIODevice::ReadOnly)) return s;
        const QJsonObject root = QJsonDocument::fromJson(f.readAll()).object();
        s.size = root.value("size").toInt(s.size);
        const QString frame = root.value("logoFrame").toString();
        if (!frame.isEmpty()) s.logoFrame = frame.at(0);
        s.palette = readPalette(root.value("palette").toObject());
        s.dark = readPalette(root.value("paletteDark").toObject());
        const QJsonObject icons = root.value("icons").toObject();
        for (auto it = icons.begin(); it != icons.end(); ++it) {
          QStringList rows;
          for (const QJsonValue& r : it.value().toArray()) rows << r.toString();
          s.icons.insert(it.key(), rows);
        }
        return s;
      }();
      return set;
    }

    // `ink` overrides single characters (the dark face, and the mark's accent ring).
    QString rectRuns(const QStringList& rows, const QHash<QChar, QString>& ink) {
      const PixelSet& s = pixelSet();
      QString out;
      for (int y = 0; y < rows.size(); ++y) {
        const QString& row = rows.at(y);
        for (int x = 0; x < row.size();) {
          const QChar ch = row.at(x);
          int w = 1;
          while (x + w < row.size() && row.at(x + w) == ch) ++w;
          const QString fill = ink.value(ch, s.palette.value(ch));
          if (!fill.isEmpty())
            out += QString(R"(<rect x="%1" y="%2" width="%3" height="1" fill="%4"/>)").arg(x).arg(y).arg(w).arg(fill);
          x += w;
        }
      }
      return out;
    }

    QHash<QChar, QString> themeInk() {
      return isWebcore() && skinDark() ? pixelSet().dark : QHash<QChar, QString>();
    }

    QString document(const QString& runs) {
      return QString(R"(<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 %1 %1" )"
                     R"(shape-rendering="crispEdges">%2</svg>)")
          .arg(pixelSet().size)
          .arg(runs);
    }
  }  // namespace

  bool hasPixelIcon(const QString& name) { return pixelSet().icons.contains(name); }

  QString pixelIconSvg(const QString& name) {
    static QHash<QString, QString> docs;
    const bool dark = isWebcore() && skinDark();
    const QString key = name + (dark ? "|d" : "|l");
    const auto hit = docs.constFind(key);
    if (hit != docs.constEnd()) return *hit;
    const PixelSet& s = pixelSet();
    const auto rows = s.icons.constFind(name);
    if (rows == s.icons.constEnd()) return QString();
    const QString doc = document(rectRuns(*rows, themeInk()));
    docs.insert(key, doc);
    return doc;
  }

  QString pixelLogoSvg(const QColor& frame) {
    const PixelSet& s = pixelSet();
    const auto rows = s.icons.constFind(QStringLiteral("logo"));
    if (rows == s.icons.constEnd()) return QString();
    QHash<QChar, QString> ink = themeInk();
    if (frame.isValid() && !s.logoFrame.isNull()) ink.insert(s.logoFrame, frame.name());
    return document(rectRuns(*rows, ink));
  }

  QImage pixelIconImage(const QString& name, bool dark, int scale) {
    const PixelSet& s = pixelSet();
    const QStringList rows = s.icons.value(name);
    if (rows.isEmpty()) return QImage();
    QImage img(rows.size() * scale, rows.size() * scale, QImage::Format_ARGB32_Premultiplied);
    img.fill(Qt::transparent);
    QPainter p(&img);
    for (int y = 0; y < rows.size(); ++y)
      for (int x = 0; x < rows.at(y).size(); ++x) {
        const QChar ch = rows.at(y).at(x);
        const QString hex = dark ? s.dark.value(ch, s.palette.value(ch)) : s.palette.value(ch);
        if (!hex.isEmpty()) p.fillRect(x * scale, y * scale, scale, scale, QColor(hex));
      }
    p.end();
    return img;
  }

}  // namespace stencil::support

File: browser/js/ui/webcore/icons.js
// The skin's icons (config/iconsWebcore.json): a palette map becomes <rect> runs on a 16-grid,
// installed into ui/icons.js so every glyph assembled afterwards wears it, and swapped into the
// ones already on the page. Desktop twin: support/webcore/icons.cpp.
import PIXELS from '../../config/iconsWebcore.json' with { type: 'json' };
import { icon, setIconSkin, DRAW_MODE_ICON } from '../icons.js';

// One <rect> per run of a lit colour along each row. `ink` re-inks single palette characters.
export const pixelRects = (rows, ink = null) => {
  let out = '';
  rows.forEach((row, y) => {
    for (let x = 0; x < row.length;) {
      const ch = row[x];
      let w = 1;
      while (x + w < row.length && row[x + w] === ch) w++;
      const fill = (ink && ink[ch]) || PIXELS.palette[ch];
      if (fill) out += `<rect x="${x}" y="${y}" width="${w}" height="1" fill="${fill}"/>`;
      x += w;
    }
  });
  return out;
};

const themeInk = (dark) => (dark ? PIXELS.paletteDark : null);   // the dark face's own ink

// The mark's ring in the preset's own hex: the skin points --accent at its navy, so no var says it.
export const pixelLogoRects = (accent, dark = false) =>
  pixelRects(PIXELS.icons.logo, { ...themeInk(dark), ...(accent ? { [PIXELS.logoFrame]: accent } : null) });

const tables = { light: null, dark: null };
export const pixelTable = (dark = false) => {
  const key = dark ? 'dark' : 'light';
  return (tables[key] ??= Object.fromEntries(
    Object.entries(PIXELS.icons).map(([n, rows]) => [n, pixelRects(rows, themeInk(dark))])));
};

// One icon as a standalone document (the favicon); `accent` inks the mark's ring.
export const pixelIconSvg = (name, size = PIXELS.size, accent = null, dark = false) => {
  const art = (name === 'logo' && accent) ? pixelLogoRects(accent, dark) : (pixelTable(dark)[name] || '');
  return `<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 16 16" width="${size}" height="${size}" ` +
    `shape-rendering="crispEdges">${art}</svg>`;
};

export const setPixelIcons = (on, dark = false) => setIconSkin(on ? pixelTable(dark) : null);

const NAME_RE = /^ic-([a-z0-9-]+)$/;
const SW_RE = /^ic-sw([\d.]+)$/;
let logoArt = null;   // the header mark's own markup and grid, kept to put back

// The header mark alone: where every accent change lands while the skin is on.
export const paintLogoMark = (root, accent, dark = false) => {
  const logo = root?.querySelector?.('svg.app-logo');
  if (logo && logo.getAttribute('viewBox') === '0 0 16 16') logo.innerHTML = pixelLogoRects(accent, dark);
};

// Redraw every glyph under `root` in whatever ui/icons.js is serving now; the header mark and the
// draw-mode faces follow, in place (their listeners stay).
export const swapIconArt = (root, on, accent = null, dark = false) => {
  for (const svg of [...root.querySelectorAll('svg.ic')]) {
    if (svg.classList.contains('draw-mode-icon')) continue;   // not in the line-art table
    const classes = [...svg.classList];
    const name = classes.map((c) => NAME_RE.exec(c)?.[1]).find(Boolean);
    if (!name) continue;
    // The line-art's stroke width rides along as a class, so the way back keeps it.
    const stroke = svg.getAttribute('stroke-width');
    const kept = classes.filter((c) => c !== 'ic' && c !== `ic-${name}` && !SW_RE.test(c));
    if (on && stroke && stroke !== '2') kept.push(`ic-sw${stroke}`);
    const sw = classes.map((c) => SW_RE.exec(c)?.[1]).find(Boolean);
    svg.outerHTML = icon(name, { size: Number(svg.getAttribute('width')) || 16, cls: kept.join(' '), sw: sw ? Number(sw) : 2 });
  }
  for (const svg of [...root.querySelectorAll('svg.draw-mode-icon')]) {
    const rect = svg.classList.contains('ic-draw-mode-rect') || /ic-box/.test(svg.innerHTML);
    svg.outerHTML = rect ? DRAW_MODE_ICON.rect : DRAW_MODE_ICON.line;
  }
  // The install button is one big arrow of its own (see iconsWebcore.json `install`).
  const install = on ? root.querySelector?.('#install-toggle svg.ic') : null;
  if (install) install.innerHTML = pixelTable(dark).install;
  const logo = root.querySelector('svg.app-logo');
  if (!logo) return;
  if (on) {
    logoArt ??= { inner: logo.innerHTML, viewBox: logo.getAttribute('viewBox') };
    logo.innerHTML = pixelLogoRects(accent, dark);
    logo.setAttribute('viewBox', '0 0 16 16');
    logo.setAttribute('shape-rendering', 'crispEdges');
  } else if (logoArt) {
    logo.innerHTML = logoArt.inner;
    logo.setAttribute('viewBox', logoArt.viewBox);
    logo.removeAttribute('shape-rendering');
  }
};
